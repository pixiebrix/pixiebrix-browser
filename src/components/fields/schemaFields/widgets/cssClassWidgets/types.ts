/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type Expression } from "@/types/runtimeTypes";
import type React from "react";

export type Spacing = {
  side: string | null;
  size: number;
};

export type Value = string | Expression;

/**
 * An independent class name
 */
export type ClassFlag = {
  /**
   * The Bootstrap 4 class name
   */
  className: string;

  /**
   * Title node to render for the element (in a button/dropdown)
   */
  title: React.ReactNode;

  /**
   * True if the flag is exclusive for it's group (default=true)
   */
  exclusive?: boolean;

  /**
   * Other flags in the same group that the flag implies
   */
  implies?: string[];
};
