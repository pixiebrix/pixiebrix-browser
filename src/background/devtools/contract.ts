/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Runtime, WebNavigation } from "webextension-polyfill-ts";
import { HandlerOptions } from "@/messaging/protocol";
import type { Target } from "@/types";

export const PORT_NAME = "devtools-page";
export const MESSAGE_PREFIX = "@@pixiebrix/devtools/";

export type TabId = number;
export type Nonce = string;

export type PromiseHandler = [
  (value: unknown) => void,
  (value: unknown) => void
];

export type BackgroundEventType =
  | "HistoryStateUpdate"
  | "DOMContentLoaded"
  | "ContentScriptReady";

export type BackgroundEvent =
  | {
      type: "HistoryStateUpdate";
      meta: Meta;
      payload: WebNavigation.OnHistoryStateUpdatedDetailsType;
    }
  | {
      type: "DOMContentLoaded";
      meta: Meta;
      payload: WebNavigation.OnDOMContentLoadedDetailsType;
    }
  | {
      type: "ContentScriptReady";
      meta: Meta;
      payload: { frameId: number; tabId: TabId };
    };

export interface BackgroundResponse {
  type: string;
  meta: Meta;
  payload: unknown;
}

export interface HandlerEntry {
  handler: (
    target: Target,
    port: Runtime.Port
  ) => (...args: unknown[]) => unknown | Promise<unknown>;
  options: HandlerOptions;
}

export interface Meta {
  nonce: Nonce;
  tabId: TabId;
  frameId: number;
  [index: string]: unknown;
}
