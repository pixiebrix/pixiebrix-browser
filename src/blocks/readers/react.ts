/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
