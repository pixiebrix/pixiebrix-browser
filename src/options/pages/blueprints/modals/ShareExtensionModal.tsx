/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./ShareExtensionModal.module.scss";

import React, { useCallback, useMemo } from "react";
import {
  Button,
  Modal,
  Col,
  Form as BootstrapForm,
  Row,
} from "react-bootstrap";
import { FormikHelpers } from "formik";
import { compact, isEmpty, pick, sortBy, uniq } from "lodash";
import { IExtension, RegistryId, UUID } from "@/core";
import * as Yup from "yup";
import { PACKAGE_REGEX } from "@/types/helpers";
import {
  appApi,
  useGetAuthQuery,
  useGetOrganizationsQuery,
} from "@/services/api";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import slugify from "slugify";
import { getLinkedApiClient } from "@/services/apiClient";
import { objToYaml } from "@/utils/objToYaml";
import { makeBlueprint } from "@/options/pages/blueprints/utils/exportBlueprint";
import { useDispatch, useSelector } from "react-redux";
import {
  RecipeDefinition,
  selectSourceRecipeMetadata,
} from "@/types/definitions";
import notify from "@/utils/notify";
import { isAxiosError } from "@/errors";
import { faGlobe, faInfoCircle } from "@fortawesome/free-solid-svg-icons";
import Form, {
  OnSubmit,
  RenderBody,
  RenderStatus,
  RenderSubmit,
} from "@/components/form/Form";
import ConnectedFieldTemplate from "@/components/form/ConnectedFieldTemplate";
import { PackageUpsertResponse } from "@/types/contract";
import extensionsSlice from "@/store/extensionsSlice";
import SwitchButtonWidget from "@/components/form/widgets/switchButton/SwitchButtonWidget";
import FieldTemplate from "@/components/form/FieldTemplate";
import { blueprintModalsSlice } from "@/options/pages/blueprints/modals/blueprintModalsSlice";
import { selectExtensions } from "@/store/extensionsSelectors";
import { RequireScope } from "@/auth/RequireScope";

const { attachExtension } = extensionsSlice.actions;

type FormState = {
  blueprintId: RegistryId;
  name: string;
  description: string;
  public: boolean;
  organizations: UUID[];
};

const ShareSchema = Yup.object().shape({
  blueprintId: Yup.string()
    .matches(PACKAGE_REGEX, "Invalid registry id")
    .required(),
  name: Yup.string().required(),
  description: Yup.string().required(),
  public: Yup.boolean().required(),
  organizations: Yup.array().of(Yup.string().required()),
});

async function convertAndShare(
  extension: IExtension,
  form: FormState
): Promise<RecipeDefinition> {
  const client = await getLinkedApiClient();

  const blueprint = makeBlueprint(extension, {
    id: form.blueprintId,
    name: form.name,
    description: form.description,
    version: "1.0.0",
  });

  const { data } = await client.post<PackageUpsertResponse>("api/bricks/", {
    config: objToYaml(blueprint),
    kind: "recipe",
    public: form.public,
    organizations: form.organizations,
    share_dependencies: true,
  });

  return {
    ...blueprint,
    sharing: pick(data, ["public", "organizations"]),
    ...pick(data, ["updated_at"]),
  };
}

// XXX: this is a connected component. To simplify testing split out the presentational component
const ShareExtensionModal: React.FC<{
  extensionId: UUID;
}> = ({ extensionId }) => {
  const extensions = useSelector(selectExtensions);
  const dispatch = useDispatch();

  const extension = useMemo(() => {
    const extension = extensions.find((x) => x.id === extensionId);
    if (extension == null) {
      throw new Error(`No persisted extension exists with id: ${extensionId}`);
    }

    return extension;
  }, [extensions, extensionId]);

  const onCancel = () => {
    dispatch(blueprintModalsSlice.actions.setShareContext(null));
  };

  // If loading the URL directly, there's a race condition if scope will be populated when the modal is mounted.
  // Not a priority to fix because user will, in general, come to the modal via the "Share" button on the main page
  const {
    data: { scope },
  } = useGetAuthQuery();
  const { data: organizations = [] } = useGetOrganizationsQuery();

  const initialValues: FormState = {
    blueprintId: compact([scope, slugify(extension.label).toLowerCase()]).join(
      "/"
    ) as RegistryId,
    name: extension.label,
    description: "Created with the PixieBrix Page Editor",
    organizations: [],
    public: false,
  };

  const handleShare: OnSubmit = useCallback(
    async (values: FormState, helpers: FormikHelpers<FormState>) => {
      try {
        const recipe: RecipeDefinition = await convertAndShare(
          extension,
          values
        );
        dispatch(
          attachExtension({
            extensionId: extension.id,
            recipeMetadata: selectSourceRecipeMetadata(recipe),
          })
        );
        notify.success("Converted/shared brick");

        // Hide the share modal and show the share link modal
        dispatch(
          blueprintModalsSlice.actions.setShareContext({
            blueprintId: recipe.metadata.id,
          })
        );

        dispatch(appApi.util.invalidateTags(["Recipes"]));
      } catch (error) {
        if (isAxiosError(error) && error.response.data.config) {
          helpers.setStatus(error.response.data.config);
          return;
        }

        notify.error({
          message: "Error converting/sharing brick",
          error,
        });
      } finally {
        helpers.setSubmitting(false);
      }
    },
    [dispatch, extension]
  );

  const renderBody: RenderBody = ({ values, setFieldValue }) => (
    <div>
      <ConnectedFieldTemplate
        name="name"
        label="Name"
        description="A name for the blueprint"
      />
      <ConnectedFieldTemplate
        name="blueprintId"
        label="Registry Id"
        description={
          <span>
            A unique id for the blueprint.{" "}
            <i>Cannot be modified once shared.</i>
          </span>
        }
      />
      <ConnectedFieldTemplate
        name="description"
        label="Description"
        description="A short description of the blueprint"
      />

      <BootstrapForm.Group as={Row}>
        <Col sm="12" className="text-info">
          <FontAwesomeIcon icon={faInfoCircle} /> The blueprint&apos;s
          dependencies will automatically also be shared.
        </Col>
      </BootstrapForm.Group>

      <ConnectedFieldTemplate
        name="public"
        as={SwitchButtonWidget}
        description={
          // \u00A0 stands for &nbsp;
          values.public ? <i>Visible to all PixieBrix users</i> : "\u00A0"
        }
        label={
          <span>
            <FontAwesomeIcon icon={faGlobe} /> Public
          </span>
        }
      />

      {sortBy(organizations, (organization) => organization.name).map(
        (organization) => {
          const checked = values.organizations.includes(organization.id);
          return (
            <FieldTemplate
              key={organization.id}
              name={organization.id}
              as={SwitchButtonWidget}
              label={organization.name}
              value={checked}
              onChange={() => {
                const next = checked
                  ? values.organizations.filter(
                      (x: string) => x !== organization.id
                    )
                  : uniq([...values.organizations, organization.id]);
                setFieldValue("organizations", next);
              }}
            />
          );
        }
      )}
    </div>
  );

  const renderSubmit: RenderSubmit = ({ isSubmitting, isValid, values }) => (
    <Modal.Footer>
      <Button variant="link" onClick={onCancel}>
        Cancel
      </Button>
      <Button type="submit" disabled={!isValid || isSubmitting}>
        {values.public || !isEmpty(values.organizations) ? "Share" : "Convert"}
      </Button>
    </Modal.Footer>
  );

  const renderStatus: RenderStatus = ({ status }) => (
    <div className="text-danger mb-3">{status}</div>
  );

  return (
    <Modal show onHide={onCancel} dialogClassName={styles.dialog}>
      <Modal.Header>
        <Modal.Title>Share as Blueprint</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <RequireScope scopeSettingsDescription="To share a blueprint, you must first set an account alias for your PixieBrix account">
          <Form
            validationSchema={ShareSchema}
            initialValues={initialValues}
            onSubmit={handleShare}
            renderStatus={renderStatus}
            renderBody={renderBody}
            renderSubmit={renderSubmit}
          />
        </RequireScope>
      </Modal.Body>
    </Modal>
  );
};

export default ShareExtensionModal;
