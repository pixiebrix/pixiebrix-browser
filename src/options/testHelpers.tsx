/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import settingsSlice from "@/store/settingsSlice";
import { blueprintModalsSlice } from "@/options/pages/blueprints/modals/blueprintModalsSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import blueprintsSlice from "@/options/pages/blueprints/blueprintsSlice";
import { recipesSlice } from "@/recipes/recipesSlice";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      settings: settingsSlice.reducer,
      options: extensionsSlice.reducer,
      blueprintModals: blueprintModalsSlice.reducer,
      blueprints: blueprintsSlice.reducer,
      recipes: recipesSlice.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
