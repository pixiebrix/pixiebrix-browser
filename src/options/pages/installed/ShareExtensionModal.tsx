/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React, { useCallback, useContext } from "react";
import {
  Button,
  Modal,
  Col,
  Form as BootstrapForm,
  Row,
} from "react-bootstrap";
import { FormikHelpers } from "formik";
import { compact, isEmpty, sortBy, uniq } from "lodash";
import { IExtension, RegistryId, UUID } from "@/core";
import * as Yup from "yup";
import { PACKAGE_REGEX } from "@/types/helpers";
import AuthContext from "@/auth/AuthContext";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import slugify from "slugify";
import { getLinkedApiClient } from "@/services/apiClient";
import { objToYaml } from "@/utils/objToYaml";
import { makeBlueprint } from "@/options/pages/installed/exportBlueprint";
import { useDispatch } from "react-redux";
import { optionsSlice } from "@/options/slices";
import { RecipeDefinition } from "@/types/definitions";
import useNotifications from "@/hooks/useNotifications";
import { push } from "connected-react-router";
import { getHumanDetail } from "@/hooks/useUserAction";
import { isAxiosError } from "@/errors";
import { faGlobe, faInfoCircle } from "@fortawesome/free-solid-svg-icons";
import Form, {
  OnSubmit,
  RenderBody,
  RenderSubmit,
} from "@/components/form/Form";
import FormikHorizontalField from "@/components/form/fields/FormikHorizontalField";
import FormikSwitchButton from "@/components/form/fields/FormikSwitchButton";
import SwitchButton from "@/components/form/fields/SwitchButton";
import { useOrganization } from "@/hooks/organization";

const { attachExtension } = optionsSlice.actions;

type FormState = {
  blueprintId: RegistryId;
  name: string;
  description: string;
  public: boolean;
  organizations: UUID[];
};

const ShareSchema = Yup.object().shape({
  blueprintId: Yup.string()
    .matches(PACKAGE_REGEX, "Invalid registry id")
    .required(),
  name: Yup.string().required(),
  description: Yup.string().required(),
  public: Yup.boolean().required(),
  organizations: Yup.array().of(Yup.string().required()),
});

async function convertAndShare(
  extension: IExtension,
  form: FormState
): Promise<RecipeDefinition> {
  const client = await getLinkedApiClient();

  const blueprint = makeBlueprint(extension, {
    id: form.blueprintId,
    name: form.name,
    description: form.description,
    version: "1.0.0",
  });

  await client.post("api/bricks/", {
    config: objToYaml(blueprint),
    kind: "recipe",
    public: form.public,
    organizations: form.organizations,
    share_dependencies: true,
  });

  return blueprint;
}

const ShareExtensionModal: React.FC<{
  extension: IExtension;
  onCancel: () => void;
}> = ({ extension, onCancel }) => {
  const notify = useNotifications();
  const dispatch = useDispatch();

  // If loading the URL directly, there's a race condition if scope will be populated when the modal is mounted.
  // Not a priority to fix because user will general come to the modal via the "Share" button on the main page
  const { scope } = useContext(AuthContext);
  const { organizations } = useOrganization();

  const initialValues: FormState = {
    blueprintId: compact([scope, slugify(extension.label).toLowerCase()]).join(
      "/"
    ) as RegistryId,
    name: extension.label,
    description: "",
    organizations: [],
    public: true,
  };

  const handleShare: OnSubmit = useCallback(
    async (values: FormState, helpers: FormikHelpers<FormState>) => {
      try {
        const recipe = await convertAndShare(extension, values);
        dispatch(
          attachExtension({
            extensionId: extension.id,
            recipeMetadata: recipe.metadata,
          })
        );
        dispatch(push("/installed"));
        notify.success("Converted/shared brick");
      } catch (error: unknown) {
        if (isAxiosError(error) && error.response.data.config) {
          helpers.setStatus(error.response.data.config);
          return;
        }

        notify.error(
          `Error converting/sharing brick: ${getHumanDetail(error)}`,
          {
            error,
          }
        );
      } finally {
        helpers.setSubmitting(false);
      }
    },
    [dispatch, notify, extension]
  );

  const renderBody: RenderBody = ({ values, setFieldValue }) => (
    <Modal.Body>
      <FormikHorizontalField
        label="Name"
        name="name"
        description="A name for the blueprint"
      />
      <FormikHorizontalField
        label="Registry Id"
        name="blueprintId"
        description={
          <span>
            A unique id for the blueprint.{" "}
            <i>Cannot be modified once shared.</i>
          </span>
        }
      />
      <FormikHorizontalField
        label="Description"
        name="description"
        description="A short description of the blueprint"
      />

      <BootstrapForm.Group as={Row}>
        <Col sm="12" className="text-info">
          <FontAwesomeIcon icon={faInfoCircle} /> The blueprint&apos;s
          dependencies will automatically also be shared.
        </Col>
      </BootstrapForm.Group>

      <FormikSwitchButton
        name="public"
        label={
          values.public ? (
            <span>
              <FontAwesomeIcon icon={faGlobe} /> Public{" "}
              <span className="text-primary">
                <i> &ndash; visible to all PixieBrix users</i>
              </span>
            </span>
          ) : (
            <span>
              <FontAwesomeIcon icon={faGlobe} /> Public
            </span>
          )
        }
      />

      {sortBy(organizations, (organization) => organization.name).map(
        (organization) => {
          const checked = values.organizations.includes(organization.id);
          return (
            <SwitchButton
              key={organization.id}
              name={organization.id}
              label={organization.name}
              value={checked}
              onChange={() => {
                const next = checked
                  ? values.organizations.filter(
                      (x: string) => x !== organization.id
                    )
                  : uniq([...values.organizations, organization.id]);
                setFieldValue("organizations", next);
              }}
            />
          );
        }
      )}
    </Modal.Body>
  );

  const renderSubmit: RenderSubmit = ({ isSubmitting, isValid, values }) => (
    <Modal.Footer>
      <Button variant="link" onClick={onCancel}>
        Cancel
      </Button>
      <Button type="submit" disabled={!isValid || isSubmitting}>
        {values.public || !isEmpty(values.organizations) ? "Share" : "Convert"}
      </Button>
    </Modal.Footer>
  );

  return (
    <Modal show onHide={onCancel}>
      <Modal.Header>
        <Modal.Title>Share as Blueprint</Modal.Title>
      </Modal.Header>
      <Form
        validateOnMount
        validationSchema={ShareSchema}
        initialValues={initialValues}
        onSubmit={handleShare}
        renderBody={renderBody}
        renderSubmit={renderSubmit}
      />
    </Modal>
  );
};

export default ShareExtensionModal;
