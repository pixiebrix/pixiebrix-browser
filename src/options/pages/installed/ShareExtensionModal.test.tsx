/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import ShareExtensionModal from "./ShareExtensionModal";
import { extensionFactory } from "@/tests/factories";
import { waitForEffect } from "@/tests/testHelpers";
import userEvent from "@testing-library/user-event";
import { Organization } from "@/types/contract";

jest.mock("@/hooks/useNotifications");
jest.mock("@/services/api", () => ({
  useGetOrganizationsQuery: () => ({ data: [] as Organization[] }),
}));

test("renders modal", async () => {
  render(
    <ShareExtensionModal
      extension={extensionFactory({
        label: "testExtension",
      })}
      onCancel={jest.fn()}
    />
  );
  await waitForEffect();
  const dialogRoot = screen.getByRole("dialog");
  expect(dialogRoot).toMatchSnapshot();
});

test("prints 'Convert' when not Public (default)", async () => {
  render(
    <ShareExtensionModal
      extension={extensionFactory({
        label: "testExtension",
      })}
      onCancel={jest.fn()}
    />
  );
  await waitForEffect();
  const dialogRoot = screen.getByRole("dialog");
  const publicSwitch = dialogRoot.querySelector(
    ".form-group:nth-child(5) .switch.btn"
  );
  expect(publicSwitch).toHaveClass("off");
  const submitButton = dialogRoot.querySelector('.btn[type="submit"]');
  expect(submitButton.textContent).toBe("Convert");
});
test("prints 'Share' when Public", async () => {
  render(
    <ShareExtensionModal
      extension={extensionFactory({
        label: "testExtension",
      })}
      onCancel={jest.fn()}
    />
  );
  await waitForEffect();
  const dialogRoot = screen.getByRole("dialog");
  const publicSwitch = dialogRoot.querySelector(
    ".form-group:nth-child(5) .switch.btn"
  );
  userEvent.click(publicSwitch);
  expect(publicSwitch).toHaveClass("on");
  const submitButton = dialogRoot.querySelector('.btn[type="submit"]');
  expect(submitButton.textContent).toBe("Share");
});
