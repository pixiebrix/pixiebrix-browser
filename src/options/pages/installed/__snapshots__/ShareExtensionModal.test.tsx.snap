// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders modal 1`] = `
<div
  aria-modal="true"
  class="fade modal show"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog"
  >
    <div
      class="modal-content"
    >
      <div
        class="modal-header"
      >
        <div
          class="modal-title h4"
        >
          Share as Blueprint
        </div>
      </div>
      <form
        class=""
        novalidate=""
      >
        <div
          class="modal-body"
        >
          <div
            class="form-group row"
          >
            <label
              class="form-label col-form-label col-sm-3"
              for="name"
            >
              Name
            </label>
            <div
              class="col-sm-9"
            >
              <input
                class="form-control"
                id="name"
                name="name"
                type="text"
                value="testExtension"
              />
              <small
                class="text-muted form-text"
              >
                A name for the blueprint
              </small>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="form-label col-form-label col-sm-3"
              for="blueprintId"
            >
              Registry Id
            </label>
            <div
              class="col-sm-9"
            >
              <input
                class="form-control"
                id="blueprintId"
                name="blueprintId"
                type="text"
                value="testextension"
              />
              <small
                class="text-muted form-text"
              >
                <span>
                  A unique id for the blueprint.
                   
                  <i>
                    Cannot be modified once shared.
                  </i>
                </span>
              </small>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <label
              class="form-label col-form-label col-sm-3"
              for="description"
            >
              Description
            </label>
            <div
              class="col-sm-9"
            >
              <input
                class="form-control is-invalid"
                id="description"
                name="description"
                type="text"
                value=""
              />
              <small
                class="text-muted form-text"
              >
                A short description of the blueprint
              </small>
            </div>
          </div>
          <div
            class="form-group row"
          >
            <div
              class="text-info col-sm-12"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-info-circle fa-w-16 "
                data-icon="info-circle"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                  fill="currentColor"
                />
              </svg>
               The blueprint's dependencies will automatically also be shared.
            </div>
          </div>
          <div
            class="form-group row"
          >
            <div
              class="col-sm-3"
            >
              <div
                class="switch btn on btn-primary"
              >
                <div
                  class="switch-group"
                >
                  <span
                    class="switch-on btn btn-primary"
                  >
                     
                  </span>
                  <span
                    class="switch-off btn btn-light"
                  >
                     
                  </span>
                  <span
                    class="switch-handle btn btn-light"
                  />
                </div>
              </div>
              <input
                checked=""
                name="public"
                type="checkbox"
              />
            </div>
            <div
              class="col-sm-9"
            >
              <span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-globe fa-w-16 "
                  data-icon="globe"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 496 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M336.5 160C322 70.7 287.8 8 248 8s-74 62.7-88.5 152h177zM152 256c0 22.2 1.2 43.5 3.3 64h185.3c2.1-20.5 3.3-41.8 3.3-64s-1.2-43.5-3.3-64H155.3c-2.1 20.5-3.3 41.8-3.3 64zm324.7-96c-28.6-67.9-86.5-120.4-158-141.6 24.4 33.8 41.2 84.7 50 141.6h108zM177.2 18.4C105.8 39.6 47.8 92.1 19.3 160h108c8.7-56.9 25.5-107.8 49.9-141.6zM487.4 192H372.7c2.1 21 3.3 42.5 3.3 64s-1.2 43-3.3 64h114.6c5.5-20.5 8.6-41.8 8.6-64s-3.1-43.5-8.5-64zM120 256c0-21.5 1.2-43 3.3-64H8.6C3.2 212.5 0 233.8 0 256s3.2 43.5 8.6 64h114.6c-2-21-3.2-42.5-3.2-64zm39.5 96c14.5 89.3 48.7 152 88.5 152s74-62.7 88.5-152h-177zm159.3 141.6c71.4-21.2 129.4-73.7 158-141.6h-108c-8.8 56.9-25.6 107.8-50 141.6zM19.3 352c28.6 67.9 86.5 120.4 158 141.6-24.4-33.8-41.2-84.7-50-141.6h-108z"
                    fill="currentColor"
                  />
                </svg>
                 Public
                 
                <span
                  class="text-primary"
                >
                  <i>
                     â€“ visible to all PixieBrix users
                  </i>
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Share
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
